sabayon-brokenlinks

```diff
@@ -0,0 +1,457 @@
+          else
+            res=$(equo q files  $pkg -q | grep $f | wc -l)
+          fi
+          if [ "$res" = 1 ] ; then
+            link2fix=true
+            echo "Link $f related to existing package. Try to reinstall $pkg."
+          else
+            MANUALLY_REMOVED_LINKS+=( "$f" )
+          fi
+
+        fi
+
+      fi
+
+    else
+      # Check if file has only one revision number (es. libudev.so.0.63)
+      res=$(perl -e "if ('$base' =~ /(.so[.][\d]+[.][\d]+$)/) { print '1' }")
+      if [ -n "$res" ] ; then
+
+        search_pkg "$base" pkg newlibdir "$lib"
+        if [ -z "$pkg" ] ; then
+
+          base=$(echo "$base" | awk 'match($0, /.so[.][0-9]+/) { print substr($0, 0, RSTART+RLENGTH-1) }')
+          search_pkg "$base" pkg newlibdir "$lib"
+          if [ -z "$pkg" ] ; then
+            log 0 "No package found for link $f."
+            MANUALLY_REMOVED_LINKS+=( "$f" )
+            return
+          fi
+
+          if [ -n "$newlibdir" ] ; then
+            log 0 "Link $base available under $newlibdir. I remove it."
+            link2remove=true
+          else
+
+            current_link=$(readlink $lib/$base)
+            log 0 "Link $f related to package $pkg (with an updated version $current_link). I remove it."
+            link2remove=true
+
+          fi
+
+        else
+
+          log 0 "Link $f of the package $pkg"
+
+          if [ -n "$newlibdir" ] ; then
+            log 0 "Link $base available under $newlibdir. I remove it."
+            link2remove=true
+          else
+
+            # Check new 
+            if [ "${GENTOO_TOOL}" = "1" ] ; then
+              res=$(qlist -IC $pkg 2>/dev/null | grep $f | wc -l)
+            else
+              res=$(equo q files  $pkg -q | grep $f | wc -l)
+            fi
+            if [ "$res" = 1 ] ; then
+              link2fix=true
+              echo "Link $f related to existing package. Try to reinstall $pkg."
+            else
+              MANUALLY_REMOVED_LINKS+=( "$f" )
+            fi
+
+          fi
+
+        fi
+
+      else
+        echo "Link $f not supported".
+        MANUALLY_REMOVED_LINKS+=( "$f" )
+
+      fi
+
+    fi
+
+  fi
+
+  if [ $link2remove = true ] ; then
+    BROKEN_LINKS2REMOVE+=( "$f" )
+  fi
+
+}
+
+rm_broken_links() {
+  for l in ${BROKEN_LINKS2REMOVE[@]} ; do
+    if [ "$DRY_RUN" = 0 ]; then
+      rm -vf ${l}
+    else
+      echo "rm -vf ${l}"
+    fi
+  done
+}
+
+print_links2remove() {
+  for l in ${MANUALLY_REMOVED_LINKS[@]} ; do
+    if [ "$FORCE_MANUAL" = 1 ] ; then
+      if [ "$DRY_RUN" = 0 ] ; then
+        rm -vf ${l}
+      else
+        echo "force: rm -vf ${l}"
+      fi
+    else
+      echo "rm -vf ${l}"
+    fi
+  done
+}
+
+parse_args() {
+
+  help_message() {
+
+    echo "Sabayon Broken Links analyzer.
+Sabayon Team (2019)
+
+$0 [opts]
+
+Available options:
+
+-h|--help         This message.
+--dry-run         Dry run execution.
+--force-manual    Force remove of links to remove manually.
```

Comment:
Can't tell what is supposed to do without looking at the code

Reply:
`--dry-run` permits to analyze directories without removing nothing. Only prints broken links.
`--force-manual`: there are two type of broken links:

a) links where it's possible to identify a package. The default logic for this are:

   - if the link is related to an existing package with the found version, just said: try to reinstall package xx
  - if the link is related to a package installed with a more upgraded version, just by default remove that link at the end of the analysis (without asking user)

b) links broken without the possibility to identify the source package that has to generate the link (in a similar way of `equo s orphans`). In this case, the default logic is: not remove the link but print at the end of the analysis the command to use for removing the file. With option `--force-manual` also this files/links are removed.

Reply:
What I mean is, I would stick with a generic `--force`, as I don't see any other `--force` already being used in another arg.

Reply:
Done ;)

Reply:
Awesome! :)

Reply:
Thanks!